File: controls/sae_2025_ws/src/integration/backend/services/wifi.py
from __future__ import annotations

import asyncio
import json
import re
import subprocess

from ..context import AppContext, TargetContext


def _target_context(
    ctx: AppContext,
    *,
    target_ctx: TargetContext | None = None,
    target_id: str | None = None,
) -> TargetContext:
    if target_ctx is not None:
        return target_ctx
    return ctx.resolve_target(target_id)


async def _policy_status(session) -> dict:
    try:
        result = await session.ssh.run(
            (
                "if command -v pennair-wifi-failover >/dev/null 2>&1; then "
                "sudo -n pennair-wifi-failover --status 2>/dev/null || "
                "pennair-wifi-failover --status 2>/dev/null; "
                "fi"
            ),
            timeout=15,
        )
        if result.returncode != 0:
            return {}
        raw = (result.stdout or "").strip()
        if not raw:
            return {}
        parsed = json.loads(raw)
        return parsed if isinstance(parsed, dict) else {}
    except Exception:
        return {}


async def wifi_status(
    ctx: AppContext,
    *,
    target_ctx: TargetContext | None = None,
    target_id: str | None = None,
) -> dict:
    session = _target_context(ctx, target_ctx=target_ctx, target_id=target_id)
    try:
        policy_status = await _policy_status(session)
        result = await session.ssh.run(
            "nmcli -f NAME,TYPE,DEVICE,STATE con show --active | tail -n +2",
            timeout=15,
        )
        if result.returncode != 0:
            return {
                "success": False,
                "error": session.ssh.friendly_error(result.stderr),
            }

        connections = []
        for line in result.stdout.strip().split("\n"):
            if not line.strip():
                continue
            parts = re.split(r"\s{2,}", line.strip())
            if len(parts) >= 4:
                connections.append(
                    {
                        "name": parts[0].strip(),
                        "type": parts[1].strip(),
                        "device": parts[2].strip(),
                        "state": parts[3].strip(),
                    }
                )

        legacy_hotspot = any(
            c["name"] == ctx.operator_config.hotspot_name for c in connections
        )
        wifi_con = next(
            (c for c in connections if c["type"] == "802-11-wireless"), None
        )
        return {
            "success": True,
            "is_hotspot": policy_status.get("is_hotspot", legacy_hotspot),
            "current_wifi": policy_status.get("current_wifi")
            or (wifi_con["name"] if wifi_con else None),
            "current_mode": policy_status.get("current_mode"),
            "effective_role": policy_status.get("effective_role"),
            "policy_source": policy_status.get("policy_source"),
            "runtime_active": policy_status.get("runtime_active"),
            "mission_started": policy_status.get("mission_started"),
            "travel_router_locked": policy_status.get("travel_router_locked"),
            "switching_disabled": policy_status.get("switching_disabled"),
            "local_ap_profile": policy_status.get("local_ap_profile"),
            "travel_router_profile": policy_status.get("travel_router_profile"),
            "allowed_ap_hosts": policy_status.get("allowed_ap_hosts") or [],
            "connections": connections,
        }
    except subprocess.TimeoutExpired:
        return {"success": False, "error": session.ssh.friendly_timeout()}
    except Exception as exc:
        return {"success": False, "error": session.ssh.friendly_error(str(exc))}


async def wifi_scan(
    ctx: AppContext,
    *,
    target_ctx: TargetContext | None = None,
    target_id: str | None = None,
) -> dict:
    session = _target_context(ctx, target_ctx=target_ctx, target_id=target_id)
    try:
        result = await session.ssh.run(
            r"nmcli -f SSID,SIGNAL,SECURITY dev wifi list --rescan yes | tail -n +2",
            timeout=20,
        )
        if result.returncode != 0:
            return {
                "success": False,
                "error": session.ssh.friendly_error(result.stderr),
                "networks": [],
            }

        networks = []
        for line in result.stdout.strip().split("\n"):
            if not line.strip():
                continue
            parts = re.split(r"\s{2,}", line.strip())
            if len(parts) >= 3:
                ssid = parts[0].strip()
                if not ssid:
                    continue
                try:
                    signal = int(parts[1].strip())
                except ValueError:
                    signal = 0
                security = parts[2].strip() if len(parts) > 2 else ""
                networks.append({"ssid": ssid, "signal": signal, "security": security})

        seen: dict[str, dict] = {}
        for network in networks:
            if (
                network["ssid"] not in seen
                or network["signal"] > seen[network["ssid"]]["signal"]
            ):
                seen[network["ssid"]] = network

        return {
            "success": True,
            "networks": sorted(seen.values(), key=lambda n: -n["signal"]),
        }
    except Exception as exc:
        return {
            "success": False,
            "error": session.ssh.friendly_error(str(exc)),
            "networks": [],
        }


async def wifi_connect(
    ctx: AppContext,
    *,
    target_ctx: TargetContext | None = None,
    target_id: str | None = None,
    ssid: str,
    password: str,
) -> dict:
    session = _target_context(ctx, target_ctx=target_ctx, target_id=target_id)
    policy_status = await _policy_status(session)
    hotspot_name = session.ssh.q(
        str(policy_status.get("local_ap_profile") or ctx.operator_config.hotspot_name)
    )
    try:
        await session.ssh.run(
            f"nmcli con down {hotspot_name} 2>/dev/null; sleep 2", timeout=15
        )
        pwd_arg = f" password {session.ssh.q(password)}" if password else ""
        result = await session.ssh.run(
            f"nmcli dev wifi connect {session.ssh.q(ssid)}{pwd_arg}", timeout=30
        )
        if result.returncode != 0:
            await session.ssh.run(f"nmcli con up {hotspot_name}", timeout=15)
            return {
                "success": False,
                "error": session.ssh.format_remote_error(
                    result.stderr, "Failed to connect"
                ),
            }

        return {"success": True, "output": f"Pi connected to {ssid}"}
    except Exception as exc:
        await session.ssh.run(f"nmcli con up {hotspot_name}", timeout=15)
        return {"success": False, "error": session.ssh.friendly_error(str(exc))}


async def wifi_hotspot(
    ctx: AppContext,
    *,
    target_ctx: TargetContext | None = None,
    target_id: str | None = None,
) -> dict:
    session = _target_context(ctx, target_ctx=target_ctx, target_id=target_id)
    policy_status = await _policy_status(session)
    hotspot_name = str(
        policy_status.get("local_ap_profile") or ctx.operator_config.hotspot_name
    )
    try:
        await session.ssh.run(
            "nmcli dev disconnect wlan0 2>/dev/null; sleep 1", timeout=15
        )
        result = await session.ssh.run(
            f"nmcli con up {session.ssh.q(hotspot_name)}",
            timeout=15,
        )
        if result.returncode != 0:
            return {
                "success": False,
                "error": session.ssh.friendly_error(result.stderr),
            }
        return {"success": True, "output": "Hotspot activated"}
    except Exception as exc:
        return {"success": False, "error": session.ssh.friendly_error(str(exc))}


async def switch_local_wifi(ssid: str, password: str) -> dict:
    try:
        cmd = ["networksetup", "-setairportnetwork", "en0", ssid]
        if password:
            cmd.append(password)
        result = await asyncio.to_thread(
            subprocess.run,
            cmd,
            capture_output=True,
            text=True,
            timeout=15,
        )
        if result.returncode != 0:
            return {
                "success": False,
                "error": result.stderr.strip() or result.stdout.strip(),
            }
        return {"success": True, "output": f"Mac connected to {ssid}"}
    except Exception as exc:
        return {"success": False, "error": str(exc)}

File: controls/sae_2025_ws/integration/packages/pi-agent/src/wifi.ts
import { execFile as execFileCallback } from "node:child_process";
import { promisify } from "node:util";
import type { WifiConnection, WifiNetwork, WifiScanResult, WifiStatus } from "@pennair/integration-core";

const execFileAsync = promisify(execFileCallback);

const HOTSPOT_NAME = process.env.HOTSPOT_NAME ?? "penn-desktop";

export interface ExecResult {
  code: number;
  stdout: string;
  stderr: string;
}

export type CommandRunner = (command: string, args: string[], timeoutMs: number) => Promise<ExecResult>;

/** Real command runner -- the default, unless a test injects a fake one. */
export const runCommand: CommandRunner = async (command, args, timeoutMs) => {
  try {
    const { stdout, stderr } = await execFileAsync(command, args, { timeout: timeoutMs });
    return { code: 0, stdout, stderr };
  } catch (error) {
    const execError = error as {
      code?: number | string;
      stdout?: string;
      stderr?: string;
      message: string;
    };
    // On a spawn failure (command not found, etc.) `code` is a string like
    // "ENOENT" and `stderr` is an empty string (not absent) -- fall back to
    // `.message` for a useful error, and normalize `code` to a number.
    const numericCode = typeof execError.code === "number" ? execError.code : -1;
    return {
      code: numericCode,
      stdout: execError.stdout ?? "",
      stderr: execError.stderr || execError.message,
    };
  }
};

/** Splits on runs of 2+ spaces, matching nmcli's fixed-width column output. */
function splitColumns(line: string): string[] {
  return line.trim().split(/\s{2,}/);
}

async function policyStatus(run: CommandRunner): Promise<Record<string, unknown>> {
  const result = await run(
    "sh",
    [
      "-c",
      "if command -v pennair-wifi-failover >/dev/null 2>&1; then " +
        "sudo -n pennair-wifi-failover --status 2>/dev/null || " +
        "pennair-wifi-failover --status 2>/dev/null; " +
        "fi",
    ],
    15_000,
  );
  const raw = result.stdout.trim();
  if (result.code !== 0 || !raw) {
    return {};
  }
  try {
    const parsed: unknown = JSON.parse(raw);
    return typeof parsed === "object" && parsed !== null ? (parsed as Record<string, unknown>) : {};
  } catch {
    return {};
  }
}

export async function wifiStatus(run: CommandRunner = runCommand): Promise<WifiStatus> {
  const policy = await policyStatus(run);
  const result = await run("nmcli", ["-f", "NAME,TYPE,DEVICE,STATE", "con", "show", "--active"], 15_000);
  if (result.code !== 0) {
    return { success: false, connections: [], error: result.stderr };
  }

  const connections: WifiConnection[] = result.stdout
    .trim()
    .split("\n")
    .slice(1) // header row
    .filter((line) => line.trim())
    .map(splitColumns)
    .filter((parts) => parts.length >= 4)
    .map(([name, type, device, state]) => ({ name, type, device, state }));

  const legacyHotspot = connections.some((c) => c.name === HOTSPOT_NAME);
  const wifiConnection = connections.find((c) => c.type === "802-11-wireless");

  return {
    success: true,
    isHotspot: typeof policy.is_hotspot === "boolean" ? policy.is_hotspot : legacyHotspot,
    currentWifi: (policy.current_wifi as string | undefined) ?? wifiConnection?.name,
    currentMode: policy.current_mode as string | undefined,
    effectiveRole: policy.effective_role as string | undefined,
    connections,
  };
}

export async function wifiScan(run: CommandRunner = runCommand): Promise<WifiScanResult> {
  const result = await run(
    "nmcli",
    ["-f", "SSID,SIGNAL,SECURITY", "dev", "wifi", "list", "--rescan", "yes"],
    20_000,
  );
  if (result.code !== 0) {
    return { success: false, networks: [], error: result.stderr };
  }

  const strongestBySsid = new Map<string, WifiNetwork>();
  for (const line of result.stdout.trim().split("\n").slice(1)) {
    if (!line.trim()) continue;
    const parts = splitColumns(line);
    if (parts.length < 3) continue;
    const [ssid, signalRaw, security] = parts;
    if (!ssid) continue;
    const signal = Number.parseInt(signalRaw, 10) || 0;
    const existing = strongestBySsid.get(ssid);
    if (!existing || signal > existing.signal) {
      strongestBySsid.set(ssid, { ssid, signal, security: security ?? "" });
    }
  }

  return {
    success: true,
    networks: [...strongestBySsid.values()].sort((a, b) => b.signal - a.signal),
  };
}

export async function wifiConnect(
  ssid: string,
  password: string,
  run: CommandRunner = runCommand,
): Promise<{ success: boolean; output?: string; error?: string }> {
  const policy = await policyStatus(run);
  const hotspotName = (policy.local_ap_profile as string | undefined) ?? HOTSPOT_NAME;

  await run("nmcli", ["con", "down", hotspotName], 15_000);

  const connectArgs = ["dev", "wifi", "connect", ssid];
  if (password) {
    connectArgs.push("password", password);
  }
  const result = await run("nmcli", connectArgs, 30_000);

  if (result.code !== 0) {
    await run("nmcli", ["con", "up", hotspotName], 15_000);
    return { success: false, error: result.stderr || "Failed to connect" };
  }

  return { success: true, output: `Pi connected to ${ssid}` };
}

export async function wifiHotspot(
  run: CommandRunner = runCommand,
): Promise<{ success: boolean; output?: string; error?: string }> {
  const policy = await policyStatus(run);
  const hotspotName = (policy.local_ap_profile as string | undefined) ?? HOTSPOT_NAME;

  await run("nmcli", ["dev", "disconnect", "wlan0"], 15_000);
  const result = await run("nmcli", ["con", "up", hotspotName], 15_000);

  if (result.code !== 0) {
    return { success: false, error: result.stderr };
  }
  return { success: true, output: "Hotspot activated" };
}
